haystack/tools/toolset.py

```diff
@@ -0,0 +1,305 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from dataclasses import dataclass, field
+from typing import Any, Dict, Iterator, List, Union
+
+from haystack.core.serialization import generate_qualified_class_name, import_class_by_name
+from haystack.tools.tool import Tool, _check_duplicate_tool_names
+
+
+@dataclass
+class Toolset:
+    """
+    A collection of related Tools that can be bootstrapped collectively and managed as a cohesive unit.
+
+    1. It provides a way to group related tools together, making it easier to manage and pass them around as a single
+       unit.
+    2. It provides a base class for implementing dynamic tool loading - by subclassing Toolset, you can create
+       custom implementations that load tools from resource endpoints (e.g., an API endpoint or a configuration file).
+       This allows you to define tools externally and load them at runtime, rather than defining each tool explicitly in
+       code.
+
+
+    Note: To implement dynamic tool loading, you must create a subclass of Toolset that implements the loading logic.
+    The base Toolset class provides the infrastructure (like registration, deduplication, and serialization), but the
+    actual loading from an endpoint must be implemented in your subclass.
+
+    Important: When implementing a custom Toolset subclass, you should perform the dynamic loading of tools in the
+    `__init__` method. This ensures tools are loaded and added when the Toolset is instantiated. The base class
+    will automatically handle tool registration and deduplication. For example:
+
+    ```python
+    class CustomToolset(Toolset):
+        def __init__(self, api_endpoint: str):
+            # Always call parent's __init__ first to initialize the tools list
+            super().__init__([])
```

Comment:
nit: why not doing the reverse? First loading tools and then creating the `Toolset`
Looks more natural to me, but maybe I'm missing something.

Reply:
Right, corrected with https://github.com/deepset-ai/haystack/pull/9161/commits/9a3b4143a02fb7fd7e17a7580bec89d3b8f6ea39